File: performance_validation/real_world/json_parser/benchmark.cpp
// JSON Parser Benchmark - C++ Implementation
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <memory>
#include <chrono>
#include <fstream>
#include <sstream>
#include <variant>
#include <stdexcept>

class JsonValue {
public:
    using Object = std::map<std::string, std::shared_ptr<JsonValue>>;
    using Array = std::vector<std::shared_ptr<JsonValue>>;
    using Value = std::variant<std::nullptr_t, bool, double, std::string, Object, Array>;
    
    Value value;
    
    JsonValue() : value(nullptr) {}
    JsonValue(bool b) : value(b) {}
    JsonValue(double d) : value(d) {}
    JsonValue(const std::string& s) : value(s) {}
    JsonValue(const Object& o) : value(o) {}
    JsonValue(const Array& a) : value(a) {}
};

class JsonParser {
private:
    std::string input;
    size_t pos = 0;
    
    void skipWhitespace() {
        while (pos < input.length() && std::isspace(input[pos])) {
            pos++;
        }
    }
    
    char peek() {
        if (pos >= input.length()) return '\0';
        return input[pos];
    }
    
    char consume() {
        if (pos >= input.length()) throw std::runtime_error("Unexpected end of input");
        return input[pos++];
    }
    
    std::string parseString() {
        consume(); // Skip opening quote
        std::string result;
        
        while (pos < input.length()) {
            char c = consume();
            if (c == '"') {
                return result;
            } else if (c == '\\') {
                char next = consume();
                switch (next) {
                    case '"': result += '"'; break;
                    case '\\': result += '\\'; break;
                    case '/': result += '/'; break;
                    case 'b': result += '\b'; break;
                    case 'f': result += '\f'; break;
                    case 'n': result += '\n'; break;
                    case 'r': result += '\r'; break;
                    case 't': result += '\t'; break;
                    case 'u': {
                        // Unicode escape
                        std::string hex;
                        for (int i = 0; i < 4; i++) {
                            hex += consume();
                        }
                        int codepoint = std::stoi(hex, nullptr, 16);
                        if (codepoint < 128) {
                            result += static_cast<char>(codepoint);
                        } else {
                            // Simplified - just add placeholder
                            result += '?';
                        }
                        break;
                    }
                    default:
                        result += next;
                }
            } else {
                result += c;
            }
        }
        
        throw std::runtime_error("Unterminated string");
    }
    
    double parseNumber() {
        size_t start = pos;
        
        if (peek() == '-') consume();
        
        // Integer part
        if (peek() == '0') {
            consume();
        } else {
            while (std::isdigit(peek())) {
                consume();
            }
        }
        
        // Fractional part
        if (peek() == '.') {
            consume();
            while (std::isdigit(peek())) {
                consume();
            }
        }
        
        // Exponent part
        if (peek() == 'e' || peek() == 'E') {
            consume();
            if (peek() == '+' || peek() == '-') {
                consume();
            }
            while (std::isdigit(peek())) {
                consume();
            }
        }
        
        std::string numStr = input.substr(start, pos - start);
        return std::stod(numStr);
    }
    
    std::shared_ptr<JsonValue> parseValue();
    
    JsonValue::Array parseArray() {
        consume(); // Skip '['
        JsonValue::Array arr;
        
        skipWhitespace();
        if (peek() == ']') {
            consume();
            return arr;
        }
        
        while (true) {
            skipWhitespace();
            arr.push_back(parseValue());
            skipWhitespace();
            
            if (peek() == ',') {
                consume();
            } else if (peek() == ']') {
                consume();
                break;
            } else {
                throw std::runtime_error("Expected ',' or ']' in array");
            }
        }
        
        return arr;
    }
    
    JsonValue::Object parseObject() {
        consume(); // Skip '{'
        JsonValue::Object obj;
        
        skipWhitespace();
        if (peek() == '}') {
            consume();
            return obj;
        }
        
        while (true) {
            skipWhitespace();
            
            if (peek() != '"') {
                throw std::runtime_error("Expected string key in object");
            }
            
            std::string key = parseString();
            skipWhitespace();
            
            if (consume() != ':') {
                throw std::runtime_error("Expected ':' after object key");
            }
            
            skipWhitespace();
            obj[key] = parseValue();
            skipWhitespace();
            
            if (peek() == ',') {
                consume();
            } else if (peek() == '}') {
                consume();
                break;
            } else {
                throw std::runtime_error("Expected ',' or '}' in object");
            }
        }
        
        return obj;
    }
    
    std::shared_ptr<JsonValue> parseValue() {
        skipWhitespace();
        
        char c = peek();
        
        if (c == '"') {
            return std::make_shared<JsonValue>(parseString());
        } else if (c == '{') {
            return std::make_shared<JsonValue>(parseObject());
        } else if (c == '[') {
            return std::make_shared<JsonValue>(parseArray());
        } else if (c == 't') {
            if (input.substr(pos, 4) == "true") {
                pos += 4;
                return std::make_shared<JsonValue>(true);
            }
        } else if (c == 'f') {
            if (input.substr(pos, 5) == "false") {
                pos += 5;
                return std::make_shared<JsonValue>(false);
            }
        } else if (c == 'n') {
            if (input.substr(pos, 4) == "null") {
                pos += 4;
                return std::make_shared<JsonValue>();
            }
        } else if (c == '-' || std::isdigit(c)) {
            return std::make_shared<JsonValue>(parseNumber());
        }
        
        throw std::runtime_error("Unexpected character in JSON");
    }
    
public:
    std::shared_ptr<JsonValue> parse(const std::string& json) {
        input = json;
        pos = 0;
        auto result = parseValue();
        skipWhitespace();
        if (pos < input.length()) {
            throw std::runtime_error("Unexpected characters after JSON value");
        }
        return result;
    }
};

// Generate test JSON
std::string generateTestJson(int depth = 3, int breadth = 4) {
    if (depth <= 0) {
        return "\"leaf\"";
    }
    
    std::stringstream ss;
    ss << "{";
    for (int i = 0; i < breadth; i++) {
        if (i > 0) ss << ",";
        ss << "\"field" << i << "\":";
        
        if (i % 3 == 0) {
            ss << "[";
            for (int j = 0; j < 3; j++) {
                if (j > 0) ss << ",";
                ss << generateTestJson(depth - 1, breadth);
            }
            ss << "]";
        } else if (i % 3 == 1) {
            ss << generateTestJson(depth - 1, breadth);
        } else {
            ss << i * 123.456;
        }
    }
    ss << "}";
    return ss.str();
}

int main(int argc, char* argv[]) {
    int iterations = argc > 1 ? std::atoi(argv[1]) : 100;
    
    // Generate complex test data
    std::string testJson = generateTestJson(5, 5);
    std::cout << "Test JSON size: " << testJson.length() << " bytes\n";
    
    // Warmup
    JsonParser warmupParser;
    for (int i = 0; i < 10; i++) {
        auto result = warmupParser.parse(testJson);
    }
    
    // Benchmark
    std::vector<double> times;
    
    for (int i = 0; i < iterations; i++) {
        auto start = std::chrono::high_resolution_clock::now();
        
        JsonParser parser;
        auto result = parser.parse(testJson);
        
        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
        times.push_back(duration.count() / 1000000.0); // Convert to seconds
    }
    
    // Calculate statistics
    double sum = 0;
    for (double t : times) sum += t;
    double avg = sum / times.size();
    
    // Output results in JSON format
    std::cout << "{\n";
    std::cout << "  \"language\": \"cpp\",\n";
    std::cout << "  \"benchmark\": \"json_parser\",\n";
    std::cout << "  \"iterations\": " << iterations << ",\n";
    std::cout << "  \"json_size\": " << testJson.length() << ",\n";
    std::cout << "  \"times\": [";
    for (size_t i = 0; i < times.size(); i++) {
        std::cout << times[i];
        if (i < times.size() - 1) std::cout << ", ";
    }
    std::cout << "],\n";
    std::cout << "  \"average_time\": " << avg << ",\n";
    std::cout << "  \"throughput_mb_per_sec\": " << (testJson.length() / avg / 1048576.0) << "\n";
    std::cout << "}\n";
    
    return 0;
}

File: performance_validation/real_world/json_parser/benchmark.rs
// JSON Parser Benchmark - Rust Implementation
use std::collections::HashMap;
use std::time::Instant;
use std::env;

#[derive(Debug, Clone)]
enum JsonValue {
    Null,
    Bool(bool),
    Number(f64),
    String(String),
    Array(Vec<JsonValue>),
    Object(HashMap<String, JsonValue>),
}

struct JsonParser {
    input: Vec<char>,
    pos: usize,
}

impl JsonParser {
    fn new(input: &str) -> Self {
        JsonParser {
            input: input.chars().collect(),
            pos: 0,
        }
    }
    
    fn skip_whitespace(&mut self) {
        while self.pos < self.input.len() && self.input[self.pos].is_whitespace() {
            self.pos += 1;
        }
    }
    
    fn peek(&self) -> Option<char> {
        if self.pos < self.input.len() {
            Some(self.input[self.pos])
        } else {
            None
        }
    }
    
    fn consume(&mut self) -> Result<char, String> {
        if self.pos >= self.input.len() {
            return Err("Unexpected end of input".to_string());
        }
        let c = self.input[self.pos];
        self.pos += 1;
        Ok(c)
    }
    
    fn parse_string(&mut self) -> Result<String, String> {
        self.consume()?; // Skip opening quote
        let mut result = String::new();
        
        while self.pos < self.input.len() {
            let c = self.consume()?;
            if c == '"' {
                return Ok(result);
            } else if c == '\\' {
                let next = self.consume()?;
                match next {
                    '"' => result.push('"'),
                    '\\' => result.push('\\'),
                    '/' => result.push('/'),
                    'b' => result.push('\u{0008}'),
                    'f' => result.push('\u{000C}'),
                    'n' => result.push('\n'),
                    'r' => result.push('\r'),
                    't' => result.push('\t'),
                    'u' => {
                        // Unicode escape
                        let mut hex = String::new();
                        for _ in 0..4 {
                            hex.push(self.consume()?);
                        }
                        if let Ok(codepoint) = u32::from_str_radix(&hex, 16) {
                            if let Some(c) = char::from_u32(codepoint) {
                                result.push(c);
                            } else {
                                result.push('?');
                            }
                        } else {
                            result.push('?');
                        }
                    }
                    _ => result.push(next),
                }
            } else {
                result.push(c);
            }
        }
        
        Err("Unterminated string".to_string())
    }
    
    fn parse_number(&mut self) -> Result<f64, String> {
        let start = self.pos;
        
        if self.peek() == Some('-') {
            self.consume()?;
        }
        
        // Integer part
        if self.peek() == Some('0') {
            self.consume()?;
        } else {
            while self.peek().map_or(false, |c| c.is_ascii_digit()) {
                self.consume()?;
            }
        }
        
        // Fractional part
        if self.peek() == Some('.') {
            self.consume()?;
            while self.peek().map_or(false, |c| c.is_ascii_digit()) {
                self.consume()?;
            }
        }
        
        // Exponent part
        if self.peek() == Some('e') || self.peek() == Some('E') {
            self.consume()?;
            if self.peek() == Some('+') || self.peek() == Some('-') {
                self.consume()?;
            }
            while self.peek().map_or(false, |c| c.is_ascii_digit()) {
                self.consume()?;
            }
        }
        
        let num_str: String = self.input[start..self.pos].iter().collect();
        num_str.parse().map_err(|e| format!("Invalid number: {}", e))
    }
    
    fn parse_array(&mut self) -> Result<Vec<JsonValue>, String> {
        self.consume()?; // Skip '['
        let mut arr = Vec::new();
        
        self.skip_whitespace();
        if self.peek() == Some(']') {
            self.consume()?;
            return Ok(arr);
        }
        
        loop {
            self.skip_whitespace();
            arr.push(self.parse_value()?);
            self.skip_whitespace();
            
            match self.peek() {
                Some(',') => {
                    self.consume()?;
                }
                Some(']') => {
                    self.consume()?;
                    break;
                }
                _ => return Err("Expected ',' or ']' in array".to_string()),
            }
        }
        
        Ok(arr)
    }
    
    fn parse_object(&mut self) -> Result<HashMap<String, JsonValue>, String> {
        self.consume()?; // Skip '{'
        let mut obj = HashMap::new();
        
        self.skip_whitespace();
        if self.peek() == Some('}') {
            self.consume()?;
            return Ok(obj);
        }
        
        loop {
            self.skip_whitespace();
            
            if self.peek() != Some('"') {
                return Err("Expected string key in object".to_string());
            }
            
            let key = self.parse_string()?;
            self.skip_whitespace();
            
            if self.consume()? != ':' {
                return Err("Expected ':' after object key".to_string());
            }
            
            self.skip_whitespace();
            let value = self.parse_value()?;
            obj.insert(key, value);
            self.skip_whitespace();
            
            match self.peek() {
                Some(',') => {
                    self.consume()?;
                }
                Some('}') => {
                    self.consume()?;
                    break;
                }
                _ => return Err("Expected ',' or '}' in object".to_string()),
            }
        }
        
        Ok(obj)
    }
    
    fn parse_value(&mut self) -> Result<JsonValue, String> {
        self.skip_whitespace();
        
        match self.peek() {
            Some('"') => Ok(JsonValue::String(self.parse_string()?)),
            Some('{') => Ok(JsonValue::Object(self.parse_object()?)),
            Some('[') => Ok(JsonValue::Array(self.parse_array()?)),
            Some('t') => {
                if self.input[self.pos..].starts_with(&['t', 'r', 'u', 'e']) {
                    self.pos += 4;
                    Ok(JsonValue::Bool(true))
                } else {
                    Err("Invalid value".to_string())
                }
            }
            Some('f') => {
                if self.input[self.pos..].starts_with(&['f', 'a', 'l', 's', 'e']) {
                    self.pos += 5;
                    Ok(JsonValue::Bool(false))
                } else {
                    Err("Invalid value".to_string())
                }
            }
            Some('n') => {
                if self.input[self.pos..].starts_with(&['n', 'u', 'l', 'l']) {
                    self.pos += 4;
                    Ok(JsonValue::Null)
                } else {
                    Err("Invalid value".to_string())
                }
            }
            Some(c) if c == '-' || c.is_ascii_digit() => {
                Ok(JsonValue::Number(self.parse_number()?))
            }
            _ => Err("Unexpected character in JSON".to_string()),
        }
    }
    
    fn parse(&mut self) -> Result<JsonValue, String> {
        let result = self.parse_value()?;
        self.skip_whitespace();
        if self.pos < self.input.len() {
            return Err("Unexpected characters after JSON value".to_string());
        }
        Ok(result)
    }
}

// Generate test JSON
fn generate_test_json(depth: i32, breadth: i32) -> String {
    if depth <= 0 {
        return "\"leaf\"".to_string();
    }
    
    let mut result = String::from("{");
    for i in 0..breadth {
        if i > 0 {
            result.push(',');
        }
        result.push_str(&format!("\"field{}\":", i));
        
        if i % 3 == 0 {
            result.push('[');
            for j in 0..3 {
                if j > 0 {
                    result.push(',');
                }
                result.push_str(&generate_test_json(depth - 1, breadth));
            }
            result.push(']');
        } else if i % 3 == 1 {
            result.push_str(&generate_test_json(depth - 1, breadth));
        } else {
            result.push_str(&format!("{}", i as f64 * 123.456));
        }
    }
    result.push('}');
    result
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let iterations = if args.len() > 1 {
        args[1].parse().unwrap_or(100)
    } else {
        100
    };
    
    // Generate complex test data
    let test_json = generate_test_json(5, 5);
    println!("Test JSON size: {} bytes", test_json.len());
    
    // Warmup
    for _ in 0..10 {
        let mut parser = JsonParser::new(&test_json);
        let _ = parser.parse();
    }
    
    // Benchmark
    let mut times = Vec::new();
    
    for _ in 0..iterations {
        let start = Instant::now();
        
        let mut parser = JsonParser::new(&test_json);
        let _ = parser.parse().unwrap();
        
        let duration = start.elapsed();
        times.push(duration.as_secs_f64());
    }
    
    // Calculate statistics
    let sum: f64 = times.iter().sum();
    let avg = sum / times.len() as f64;
    
    // Output results in JSON format
    println!("{{");
    println!("  \"language\": \"rust\",");
    println!("  \"benchmark\": \"json_parser\",");
    println!("  \"iterations\": {},", iterations);
    println!("  \"json_size\": {},", test_json.len());
    print!("  \"times\": [");
    for (i, time) in times.iter().enumerate() {
        print!("{}", time);
        if i < times.len() - 1 {
            print!(", ");
        }
    }
    println!("],");
    println!("  \"average_time\": {},", avg);
    println!("  \"throughput_mb_per_sec\": {}", test_json.len() as f64 / avg / 1048576.0);
    println!("}}")
}